pkg/common/context.go

```diff
@@ -10,3 +10,15 @@ func IsDone(ctx context.Context) bool {
 		return false
 	}
 }
+
+// CancellableWrite blocks on writing the item to the channel but can be
+// cancelled by the context. If both the context is cancelled and the channel
+// write would succeed, either operation will be performed randomly.
+func CancellableWrite[T any](ctx context.Context, ch chan<- T, item T) error {
```

Comment:
question: Does it make sense to use maybe a custom type instead of any for the the chan. I think that added type security will save us from potentially putting something of the wrong type into the channel. I'm sure the receiver of the item will likely check the type, but this way we can hopefully catch it at compile time.

Reply:
I would agree with that, except this is a generic function and the `any` here is a [type constraint](https://go.dev/doc/tutorial/generics#add_generic_function), which means it works with a channel of any type `T`. Additionally, the `item` must also be *the same* `T`.

Calling this function with `chan int` and `string` will not compile.